### PR TITLE
Fix intersections of constant arrays

### PR DESCRIPTION
**Pull request by MidnightDesign**
Allows intersection of constant arrays (`array{key: type}`)

`array{foo: string} & array{bar: int}` => `array{foo: string, bar: int}`
`array{foo?: string} & array{foo: string}` => `array{foo: string}`
`array{foo: string} & array{foo: int}` => `array{foo: never}`

To do/check:

- [x] What about `array{foo: array{bar: string}} & array{foo: array{baz: int}}`? [TypeScript version](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKAzBAuNUBGAhgE64DOwxAlgHYDmUAvk1AGR5Zy7pEBeuNAK4BbfBGItGAbgBQMgMZwaFTDWRQAFAA9c8OAEpkAPjQyo5qMQjBBxNQAMAJKi0A6Tq6LFmztx76M9rKMchg0Guic3AQkuADkcQA0MfxQAAyS+rJhEZg4eF7xcZnZ4ZH5PISpGcyMWTI55Vx4tfWNec2CNAAmEBi0EN1MbWV1UkA)

### CODE REVIEW

File: src/Type/Constant/ConstantArrayType.php [hunk 1]
```diff
@@ -58,7 +61,7 @@ class ConstantArrayType extends ArrayType implements ConstantType
 	 * @param array<int, ConstantIntegerType|ConstantStringType> $keyTypes
 	 * @param array<int, Type> $valueTypes
 	 * @param int $nextAutoIndex
-	 * @param int[] $optionalKeys
+	 * @param list<int> $optionalKeys
```
Comment (MidnightDesign) [inline src/Type/Constant/ConstantArrayType.php:64]: I guess this is _technically_ a BC break, but I can't see why anybody would have passed anything other than a list here.

**Comment by MidnightDesign** [comment]
@ondrejmirtes I can't think of any more (edge) cases, can you?

Even if I missed something, it's probably better than not working at all.

**Comment by MidnightDesign** [comment]
I'm not sure what to do about the failing `testIntersectInversed` cases since the expected order of the keys is the order of the arrays. We could normalize the keys to always be in alphabetical order, but that might be a bit surprising to users.

I guess the best option would be to just normalize the order just inside `testIntersectInversed()`.

**Comment by ondrejmirtes** [comment]
Hi, sorry, I don't see this PR ever to make it over. `ConstantArrayType` right now representes both PHPDoc array shapes and literal arrays in code. And intersection of literal arrays needs to work differently in code than you want it to work in PHPDocs.

The distinction here is that we work with types vs. values, but unfortunately ConstantArrayType represents both.

**Comment by MidnightDesign** [comment]
I can't see the difference. The value `['foo' => 42]` is of type `array{foo: 42}` (or `array{foo: int}` if you want to generalize it like TypeScript does - maybe to save memory?). But if you have other plans to support intersections of `array{...}` types, I'm all for it.

**Pull request closed without merging**